File: local_server.js
import http from 'http';
import url from 'url';
import * as dotenv from 'dotenv';
import handler from './api/horoscope-api.js';

// Carregar variáveis de ambiente
dotenv.config();

// Porta na qual o servidor vai rodar
const PORT = 9999;

// Servidor HTTP simples
const server = http.createServer((req, res) => {
  // Configurar CORS para todas as respostas
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  
  // Tratamento de requisições OPTIONS (preflight CORS)
  if (req.method === 'OPTIONS') {
    res.statusCode = 200;
    res.end();
    return;
  }
  
  // Verificar se é uma requisição POST para o endpoint correto
  if (req.method === 'POST' && req.url === '/api/horoscope') {
    let body = '';
    
    // Coletar dados da requisição
    req.on('data', chunk => {
      body += chunk.toString();
    });
    
    // Processar os dados quando a requisição terminar
    req.on('end', async () => {
      try {
        // Converter os dados para objeto
        const data = JSON.parse(body);
        
        // Criar objeto de requisição e resposta adaptados para o handler
        const mockReq = {
          method: 'POST',
          body: data
        };
        
        const mockRes = {
          statusCode: 200,
          headers: {},
          body: '',
          
          // Métodos para simular a API do objeto res do Vercel
          status(code) {
            this.statusCode = code;
            return this;
          },
          
          setHeader(name, value) {
            this.headers[name] = value;
          },
          
          json(data) {
            this.body = JSON.stringify(data);
            return this;
          },
          
          end() {
            return this;
          }
        };
        
        // Chamar o handler da API
        await handler(mockReq, mockRes);
        
        // Configurar a resposta HTTP
        res.statusCode = mockRes.statusCode;
        
        // Adicionar cabeçalhos
        Object.keys(mockRes.headers).forEach(key => {
          res.setHeader(key, mockRes.headers[key]);
        });
        
        // Sempre enviar como JSON
        res.setHeader('Content-Type', 'application/json');
        
        // Enviar resposta
        res.end(mockRes.body);
      } catch (error) {
        console.error('Erro ao processar requisição:', error);
        res.statusCode = 500;
        res.setHeader('Content-Type', 'application/json');
        res.end(JSON.stringify({
          error: 'Erro interno do servidor',
          message: error.message
        }));
      }
    });
  } else {
    // Para outros caminhos ou métodos, retornar 404
    res.statusCode = 404;
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify({ error: 'Caminho não encontrado' }));
  }
});

// Iniciar o servidor
server.listen(PORT, () => {
  console.log(`Servidor rodando na porta ${PORT}`);
  console.log(`Endpoint disponível:`);
  console.log(`  - http://localhost:${PORT}/api/horoscope`);
}); 

File: local_server.py
import json
import http.server
import socketserver
from urllib.parse import urlparse, parse_qs
from api.horoscope import handler

# Porta na qual o servidor vai rodar
PORT = 9999

class HoroscopeServer(http.server.SimpleHTTPRequestHandler):
    """Servidor simples para fornecer a função de horóscopo como um endpoint HTTP"""
    
    def do_OPTIONS(self):
        """Manipular solicitações OPTIONS para CORS"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        self.send_header('Content-Length', '0')
        self.end_headers()
    
    def do_POST(self):
        """Manipular solicitações POST"""
        # Verificar o caminho da URL
        if self.path == '/api/horoscope' or self.path == '/api/horoscope.py':
            # Obter dados da solicitação
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            
            # Converter dados para dicionário
            body = json.loads(post_data.decode('utf-8'))
            
            # Criar um evento simulado para a função
            mock_event = {
                'httpMethod': 'POST',
                'body': json.dumps(body)
            }
            
            # Chamar a função do horóscopo
            result = handler(mock_event, {})
            
            # Obter o código de status da resposta
            status_code = result.get('statusCode', 500)
            
            # Configurar a resposta HTTP
            self.send_response(status_code)
            
            # Adicionar cabeçalhos CORS
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
            
            # Encerrar cabeçalhos
            self.end_headers()
            
            # Enviar resposta
            self.wfile.write(result.get('body', '').encode('utf-8'))
        else:
            # Para outros caminhos, retornar 404
            self.send_response(404)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Caminho não encontrado'}).encode('utf-8'))

# Configurar e iniciar o servidor
def start_server():
    with socketserver.TCPServer(("", PORT), HoroscopeServer) as httpd:
        print(f"Servidor rodando na porta {PORT}")
        print(f"Endpoints disponíveis:")
        print(f"  - http://localhost:{PORT}/api/horoscope")
        print(f"  - http://localhost:{PORT}/api/horoscope.py")
        httpd.serve_forever()

if __name__ == "__main__":
    start_server() 